File: C# Unity Build/Dragon_Raid/Assets/Scripts/PlayerController.cs
using UnityEngine;
using UnityEngine.UI;
using TMPro; 

public class PlayerController : MonoBehaviour
{
    public GameObject actionCanvas;
    public GameObject skillCanvas;
    
    [HideInInspector]
    public CombatManager combatManager; 

    [Header("Player Stats")]
    public int maxHealth = 50; 
    public int currentHealth;
    public int maxAP = 50;     
    public int currentAP;      
    public int attack = 10;  
    public int defense = 5; 
    public int luck = 5;     

    [Header("UI References")]
    public TextMeshProUGUI healthText; 
    public TextMeshProUGUI apText;     


    void Start()
    {
        actionCanvas.GetComponent<Canvas>().worldCamera = Camera.main;
        currentHealth = maxHealth;
        currentAP = maxAP; 
        
        UpdateUI(); // Set the initial values
    }
    
    /**
     * A single function to update all our UI elements
     */
    public void UpdateUI()
    {
        // Update the Health Bar
        if (healthText != null && combatManager != null)
        {
            healthText.text = GenerateBarString("HP", currentHealth, maxHealth);
        }

        // Update the AP Bar
        if (apText != null)
        {
            apText.text = GenerateBarString("AP", currentAP, maxAP);
        }
    }


    public string GenerateBarString(string label, int current, int max)
    {
        int barLength = 10;
        char fillChar = '█';
        char emptyChar = '░';

        // Safety check to prevent divide-by-zero
        if (max == 0) return $"{label}: ERROR";

        // Clamp current value just in case
        current = Mathf.Clamp(current, 0, max);
        
        // Calculate the percentage and fill amount
        float percent = (float)current / max;
        int fillCount = Mathf.RoundToInt(percent * barLength);
        int emptyCount = barLength - fillCount;

        // Build the strings
        string fillString = new string(fillChar, fillCount);
        string emptyString = new string(emptyChar, emptyCount);

        // Return the final formatted string
        return label + ": [" + fillString + emptyString + "]";
        
    }


    // This is called by your buttons in the Inspector
    public void UseSkill(SkillData skill)
    {
        if (combatManager != null && combatManager.state == CombatState.PLAYERTURN)
        {
            if (currentAP >= skill.apCost)
            {
                combatManager.OnPlayerSkill(skill, attack, luck);
                currentAP -= skill.apCost;
            }
            else
            {
                Debug.Log("Not enough AP to use " + skill.skillName);
            }
        }
    }

    public void OnCycleTarget()
    {
        if (combatManager != null && combatManager.state == CombatState.PLAYERTURN)
        {
            combatManager.CycleTarget();
        }
    }

    public void TakeDamage(int attackerAttack, int attackerLuck, int attackerPower)
    {
        float randMin = combatManager.randMinMultiplier;
        float randMax = combatManager.randMaxMultiplier;
        float atkPotency = combatManager.attackPotency;
        float defDivisor = combatManager.defenseDivisor;
        float defPotency = combatManager.defensePotency;
        float defPenalty = combatManager.defensePenaltyMultiplier;
        float defScalar = combatManager.defensePenaltyScalar;
        float defToggle = combatManager.defensePenaltyToggle;
        float flatBonus = combatManager.flatDamageBonus;

        float critChancePerLuck = combatManager.lucktoCrit;
        float critMultiplier = combatManager.critDamageMultiplier;

        // Formula Breakdown:
        float ATK = attackerAttack;
        float PWR = attackerPower;
        float DEF = this.defense; 

        float randValue = Random.Range(randMin * PWR, randMax * PWR);

        if (defDivisor == 0) defDivisor = 1f;
        
        float mainMultiplier = (ATK * atkPotency) / (DEF * defDivisor);
        float penalty = (DEF * defPotency * defPenalty) * defScalar * defToggle;

        int baseDamage = Mathf.FloorToInt(randValue * (mainMultiplier - penalty) + flatBonus);

        float critChance = attackerLuck * critChancePerLuck;
        float critRoll = Random.Range(0f, 100f);

        int finalDamage;
        if (critRoll < critChance)
        {
            finalDamage = (int)(Mathf.FloorToInt(baseDamage * critMultiplier));
            Debug.Log("Critical Hit!");
        }
        else
        {
            finalDamage = baseDamage;
        }
        
        if (finalDamage < 1) finalDamage = 1;
    
        currentHealth -= finalDamage;
        Debug.Log("Player takes " + finalDamage + " damage!");

        if (currentHealth <= 0)
        {
            Die();
        }
    }

    void Die()
    {
        Debug.Log("Player has been defeated!");
        combatManager.state = CombatState.LOSE;
    }
    
    public void SetActionPanel(bool show)
    {
        if (actionCanvas != null) actionCanvas.SetActive(show);
    }

    public void SetSkillPanel(bool show)
    {
        if (skillCanvas != null) skillCanvas.SetActive(show);
    }
}

File: DragonRR/DragonRR/src/commonMain/kotlin/samples/mygame/player/playerController.kt
package com.yourgame.player
import com.soywiz.korge.view.*
import com.soywiz.kmmuix.ui.text.Text
import com.soywiz.korge.scene.Scene
import kotlin.random.Random

/**
 *player controller class for korGE
 */
class PlayerController(
    val views: Views,  //korGE Views context
    var combatManager: CombatManager? = null  //will be set from scene
) {

    //player stats
    var maxHealth: Int = 50
    var currentHealth: Int = 50
    var maxAP: Int = 50
    var currentAP: Int = 50
    var attack: Int = 10
    var defense: Int = 5
    var luck: Int = 5

    // UI References - KorGE Text views instead of TextMeshProUGUI
    var healthText: Text? = null
    var apText: Text? = null

    suspend fun init() {
        currentHealth = maxHealth
        currentAP = maxAP
        updateUI()
    }
    /**
     * A single function to update all our UI elements
     */
    fun updateUI() {
        // Update the Health Bar
        healthText?.text = generateBarString("HP", currentHealth, maxHealth)
        apText?.text = generateBarString("AP", currentAP, maxAP)
    }
    fun generateBarString(label: String, current: Int, max: Int): String {
        val barLength = 10
        val fillChar = '█'
        val emptyChar = '░'

        //safety check to prevent divide-by-zero
        if (max == 0) return "$label: ERROR"

        //clamp current value just in case
        val clampedCurrent = kotlin.math.minOf(current, max).coerceAtLeast(0)

        //calculate the percentage and fill amount
        val percent = clampedCurrent.toFloat() / max
        val fillCount = kotlin.math.roundToInt(percent * barLength)
        val emptyCount = barLength - fillCount

        //build the strings
        val fillString = fillChar.toString().repeat(fillCount)
        val emptyString = emptyChar.toString().repeat(emptyCount)

        //return the final formatted string
        return "$label: [$fillString$emptyString]"
    }

    //this is called by your buttons in KorGE
    fun useSkill(skill: SkillData) {
        val manager = combatManager ?: return
        if (manager.state == CombatState.PLAYERTURN && currentAP >= skill.apCost) {
            manager.onPlayerSkill(skill, attack, luck)
        } else {
            println("Not enough AP to use ${skill.skillName}")
        }
    }

    fun onCycleTarget() {
        combatManager?.let { manager ->
            if (manager.state == CombatState.PLAYERTURN) {
                manager.cycleTarget()
            }
        }
    }
    fun takeDamage(attackerAttack: Int, attackerLuck: Int, attackerPower: Int) {
        val manager = combatManager ?: return

        val randMin = manager.randMinMultiplier
        val randMax = manager.randMaxMultiplier
        val atkPotency = manager.attackPotency
        val defDivisor = manager.defenseDivisor.coerceAtLeast(1f)
        val defPotency = manager.defensePotency
        val defPenalty = manager.defensePenaltyMultiplier
        val defScalar = manager.defensePenaltyScalar
        val defToggle = manager.defensePenaltyToggle
        val flatBonus = manager.flatDamageBonus

        val critChancePerLuck = manager.lucktoCrit
        val critMultiplier = manager.critDamageMultiplier

        // Formula Breakdown:
        val ATK = attackerAttack
        val PWR = attackerPower
        val DEF = defense

        val randValue = Random.nextFloat() * (randMax - randMin) * PWR + randMin * PWR

        val mainMultiplier = (ATK * atkPotency) / (DEF * defDivisor)
        val penalty = (DEF * defPotency * defPenalty) * defScalar * defToggle

        val baseDamage = kotlin.math.floor(randValue * (mainMultiplier - penalty) + flatBonus).toInt()

        val critChance = attackerLuck * critChancePerLuck
        val critRoll = Random.nextFloat() * 100f

        val finalDamage = if (critRoll < critChance) {
            println("Critical Hit!")
            kotlin.math.floor(baseDamage * critMultiplier).toInt()
        } else {
            baseDamage
        }.coerceAtLeast(1)

        currentHealth -= finalDamage
        println("Player takes $finalDamage damage!")

        if (currentHealth <= 0) {
            die()
        }
    }
    private fun die() {
        println("Player has been defeated!")
        combatManager?.state = CombatState.LOSE
    }
    fun setActionPanel(show: Boolean) {
        // In KorGE: toggle visibility of your action UI container View
        // actionContainer?.visible = show
    }
    fun setSkillPanel(show: Boolean) {
        // in KorGE: toggle visibility of your skill UI container View
        // skillContainer?.visible = show
    }
}
